File: Cancer/old/Get_Elliptical_Area.java
import ij.*;
import ij.process.*;
import ij.gui.*;
import ij.plugin.*;
import ij.plugin.frame.*;
import ij.plugin.filter.*;
import ij.measure.*;
import ij.io.*;

import java.io.IOException;
import loci.formats.FormatException;
import loci.plugins.BF;

import java.awt.*;
import java.lang.Math;
import java.util.regex.*;
import java.io.File;
import java.io.IOException;


/**
 * All units are in mm except when otherwise noted.
 */
public class Get_Elliptical_Area implements PlugIn {

	//***** Variables to Set *******
	private double minimumCancerRadius = 0.3;
	private double backgroundRadius = 30;
	private double medianRadius = 5;
	private double lineWidth = 10;
	private double closeNum = 5;

	private String queryDirectory = "/tmp/queried_images";
	
	//***** Everything Else ********
	private double minimumCancerArea = Math.PI * Math.pow(minimumCancerRadius, 2);

	private ResultsTable results = new ResultsTable();
	private Analyzer analyzer;

	private int fileCounter = 0;
	
	public void run(String arg) {
		//***** Set Correct Measurements ******
		initialize_Analyzer();
		
		//***** Get Input and Output Directories *****
		
		// Iterate over all Input Images
		File folder = new File(queryDirectory);
		File[] listOfFiles = folder.listFiles();
		for (int i = 0; i < listOfFiles.length; i++) {
			File current_file = listOfFiles[i];
			IJ.showProgress(i, listOfFiles.length);
			if (current_file.isFile()) {
				try {					
					//*****Open the File*****
					ImagePlus images[] = BF.openImagePlus(current_file.getPath());
					ImagePlus originalImage = images[0];
					
					//Resize the image
					String resizeTitle = "scaledImage.tiff";
					IJ.run(originalImage, "Scale...", "x=0.25 y=0.25 width=347 height=260 interpolation=Bicubic average process create title=" + resizeTitle);
					originalImage = WindowManager.getImage(resizeTitle);
					//Make 8-bit
					IJ.run(originalImage, "8-bit", "");
					
					//originalImage.show();
					
					ImagePlus thresholdedImage = originalImage.duplicate();
					thresholdedImage.show();
					findTumor(thresholdedImage);
					measureTumor(thresholdedImage);
					// Create an overlay
					ImagePlus overlayImage = createOverlay(originalImage, thresholdedImage);
					
					saveResults();
					
				}
				catch (FormatException exc) {
					IJ.error("Potato: " + exc.getMessage());
				}
				catch (IOException exc) {
					IJ.error("POTATO: " + exc.getMessage());
				}
			}
		}
	}

	public String getPlateIdentity(String fileName) {
		// Get Plate identity
		Pattern pattern = Pattern.compile("_[A-Z][0-9][0-9]_");
		Matcher matcher = pattern.matcher(fileName);
		matcher.find();
		String name = matcher.group();
		name = name.substring(1, 4);
		return name;
	}
	
	private void initialize_Analyzer() {
		analyzer = new Analyzer();
		analyzer.setResultsTable(results);
		analyzer.setMeasurement(analyzer.AREA, true);
		analyzer.setMeasurement(analyzer.MEAN, true);
		analyzer.setMeasurement(analyzer.CIRCULARITY, true);
		analyzer.setMeasurement(analyzer.ELLIPSE, true);
		analyzer.setMeasurement(analyzer.SHAPE_DESCRIPTORS, true);
		analyzer.setMeasurement(analyzer.LABELS, true);
		analyzer.setMeasurement(analyzer.KURTOSIS, true);
		analyzer.setPrecision(3);
	}
	/**
	 * Responsible for finding the tumor in the image passed in. Works on stacks.
	 */
	public void findTumor(ImagePlus image){
		// Determine if the tumor is black on a white background or white on a black background
		// based on the mean gray value
		ImageStatistics stats = image.getStatistics(Measurements.MEAN);
		IJ.log(Double.toString(stats.mean));
		
		IJ.run(image, "Subtract Background...", "rolling=" + String.valueOf(backgroundRadius) + " sliding stack");
		IJ.run(image, "Auto Threshold", "method=MaxEntropy white");
		// Set background to white so things are removed correctly
		IJ.setForegroundColor(0, 0, 0);
		IJ.run(image, "Particle Remover edited", "size=0-" + String.valueOf(minimumCancerArea)  
			+ " circularity=0.00-1.00 show=Nothing include");
		IJ.run(image, "Options...", "iterations=" + String.valueOf(closeNum) + " count=1 black edm=Overwrite do=Close");
		IJ.runPlugIn(image, "Binary", "Fill Holes");
		IJ.runPlugIn(image, "Binary", "Watershed");
		IJ.run(image, "Particle Remover edited", "size=0-" + String.valueOf(minimumCancerArea)  
			+ " circularity=0.00-1.00 show=Nothing include");
	}
	
	public void measureTumor(ImagePlus image) {
		IJ.run(image, "Analyze Particles...", "size=0-Infinity circularity=0.00-1.00 show=Nothing display include");
	}

	public ImagePlus createOverlay(ImagePlus original, ImagePlus thresholded) {
		ImagePlus overlayImage = original.duplicate();
		//Locate the particles
		IJ.run(thresholded, "Analyze Particles...", "size=0-Infinity circularity=0.00-1.00 show=Nothing add");
		//Prepare Image
		IJ.runPlugIn(overlayImage, "Converter", "RGB Color");
		IJ.setForegroundColor(255, 201, 0);
		IJ.setBackgroundColor(0, 0, 0);
		IJ.run("Line Width...", "line=" + lineWidth);
		IJ.runPlugIn("roiManager", "Label");
		
		return overlayImage;
	}
	
	public void saveResults() {
		
	}
}

File: Cancer/getEllipticalArea_.py
from ij import IJ
from ij.plugin.filter import ParticleAnalyzer
from ij.plugin.frame import RoiManager
from ij.measure import Calibration
from loci.plugins import BF
from ij.io import FileSaver
from ij.measure import Measurements
from ij.process import ImageStatistics as IS
from java.awt import Color
# Python imports
import os
import math
import sys
import socket

#### Variables to set ####
hostname = socket.gethostname()
print hostname
inputDirectory = None
thumbnailDirectory = None
if hostname == 'dawnofhope':
	inputDirectory = '/home/bryan/Desktop/queried_images'
	thumbnailDirectory = '/home/bryan/Desktop/queried_images/thumbnails'
elif hostname == 'Chompert':
	inputDirectory = 'C:/Users/chompy_local/Desktop/queried_images'
	thumbnailDirectory = 'C:/Users/chompy_local/Desktop/queried_images/thumbnails'
else:
	inputDirectory = '/tmp/queried_images'
	thumbnailDirectory = '/home/bweinstein/apps/OMERO/thumbnails'
outputFile = 'Areas.txt' # Goes in the input directory
minimumCancerArea = 0.03
maxCancerArea = 1
circularityCutoff = .65 # Dimensionless
backgroundRadius = 30 # in pixels
medianSmoothing = 5 # in pixels

############# Auxillary Functions #########
def locateTumors(imp):
	'''Locates the tumors in an image; white is location, black is not'''
	# Determine if the tumor is black on a white background or white
	# on a black background
	ip = imp.getProcessor()
	stats = IS.getStatistics(ip, IS.MEAN, imp.getCalibration())
	mean = stats.mean
	if mean >= 20: # Black tumor on white background
		IJ.run(imp, 'Subtract Background...', 'rolling=' + str(backgroundRadius) + ' light sliding stack');
		IJ.run(imp, "Invert", "stack")
	else:
		IJ.run(imp, 'Subtract Background...', 'rolling=' + str(backgroundRadius) + ' sliding stack');
	IJ.run(imp, 'Median...', 'radius=' + str(medianSmoothing))
	IJ.run(imp, 'Auto Threshold', 'method=MaxEntropy white stack')
	# Remove small
	IJ.setForegroundColor(0, 0, 0);
	IJ.run(imp, 'ParticleRemoverPy ', 'enter=' + str(minimumCancerArea));
	IJ.run(imp, "Close-", "stack")
	# Make sure black background binary is set!
	IJ.run(imp, "Options...", "iterations=1 count=1 black edm=Overwrite");
	IJ.run(imp, 'Fill Holes', "stack")
	IJ.run(imp, 'Watershed', "stack")
	IJ.run(imp, 'ParticleRemoverPy ', 'enter=' + str(minimumCancerArea));
	
def measureTumor(original, locations):
	'''Returns the area from the original image with the 
	highest kurtosis which generally corresponds to the most
	in focus image. Also saves an image corresponding to a mask
	of the measurement.'''
	# Prevent ROI manager from appearing
	roiM = RoiManager(True)
	ParticleAnalyzer.setRoiManager(roiM)
	# Locate particles above a minimum size and with a desired circularity
	IJ.run(locations, "Analyze Particles...", "size=" + str(minimumCancerArea) +"-" + str(maxCancerArea) +" circularity=" + str(circularityCutoff) + "-1.00 show=Nothing exclude add stack");
	# Choose ROI with the highest kurtosis
	maxKurtosis = None
	area = None
	selectedROI = None
	for roi in roiM.getRoisAsArray():
		original.setRoi(roi)
		stats = original.getStatistics(Measurements.KURTOSIS, Measurements.AREA)
		currentKurtosis = stats.kurtosis
		if currentKurtosis > maxKurtosis:
			maxKurtosis = stats.kurtosis
			area = stats.area
			selectedROI = roi
	original.killRoi() # Remove the remaining ROI
	roiM.runCommand("Reset")
	return (area, selectedROI)

def createOverlay(image, selectedROI):
	overlayImage = image.duplicate()
	# Convert to RGB
	IJ.run(overlayImage, "RGB Color", "")
	overlayImage.setOverlay(selectedROI, Color(246, 27, 27), 3, None)
	roiSliceNumber = selectedROI.getPosition()
	totalNumSlices = overlayImage.getStackSize()
	count = 0
	roiPassed = False
	for i in range(totalNumSlices):
		count += 1
		if not roiPassed:
			overlayImage.setSlice(1)
			if count == roiSliceNumber:
				roiPassed = True
			else:
				IJ.run(overlayImage, "Delete Slice", "")
		else:
			overlayImage.setSlice(2)
			IJ.run(overlayImage, "Delete Slice", "")
	return overlayImage

############## Main Process ###############

# Check if output file exists
outputFilePath = inputDirectory + '/' + outputFile
if os.path.isfile(outputFilePath):
	os.remove(outputFilePath)
outputFile = open(outputFilePath, 'a')

for fileName in os.listdir(inputDirectory):
	print 'Processing Image ' , fileName
	if '.tif' in fileName:
		#print 'Opening image...'
		imps = BF.openImagePlus(os.path.join(inputDirectory, fileName))
		originalImage = imps[0]
	
		# Resize the image
		numSlices = originalImage.getStackSize()
		IJ.run(originalImage, "Size...", "width=347 height=260 depth=" + str(numSlices) +" constrain average interpolation=Bicubic");
		
		# Make 8 bit
		IJ.run(originalImage, '8-bit', '')
		tumorLocations = originalImage.duplicate()
		print 'Locating tumors...'
		locateTumors(tumorLocations)
		(area, selectedROI) = measureTumor(originalImage, tumorLocations)
		print 'Area:' , area
		print 'Creating thumbnail of overlay:'
		if selectedROI is not None:
			overlayImage = createOverlay(originalImage, selectedROI)
		else:
			overlayImage = originalImage.duplicate() # I didn't feel like deleting the stack again
		overlayFileName = thumbnailDirectory + '/' + originalImage.getShortTitle() + '.png'
		FileSaver(overlayImage).saveAsPng(overlayFileName)
		# Finally write what the area is
		outputFile.write(originalImage.getTitle() + '\t' + str(area) + '\n')
# Close the original file
outputFile.close()
print 'Done!'